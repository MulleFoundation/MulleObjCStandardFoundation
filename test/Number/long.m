#ifdef __MULLE_OBJC__
# import <MulleObjCValueFoundation/MulleObjCValueFoundation.h>
#else
# import <Foundation/Foundation.h>
#endif

#include <stdio.h>


//#define PRINT_ENCODE
//#define PRINT_CLASS

static int   fail( char *title, NSNumber *a, NSNumber *b, int index)
{
   printf( "%d: %s Number FAIL %s vs. %s",
         index, title,
         [[a description] UTF8String],
         [[b description] UTF8String]);
#ifdef PRINT_ENCODE
   printf( " (%s vs. %s)", [a objCType], [b objCType]);
#endif
#ifdef PRINT_ENCODE
   printf( " (%s vs. %s)", [NSStringFromClass( [a class]) UTF8String],
                           [NSStringFromClass( [b class]) UTF8String]);
#endif
   printf( "\n");
   return( 1);
}


static int   check( char *title, NSNumber *a, NSNumber *b, int *index)
{
   int   fails;

   fails = 0;
   if( ! [a isEqualToNumber:b])
      fails += fail( title, a, b, *index);
   ++*index;

   if( ! [a isEqual:b])
      fails += fail( title, a, b, *index);
   ++*index;

   return( fails);
}


static void   test( long value)
{
   int       index = 1;
   NSNumber  *nr;
   int       fails;

   nr = [NSNumber numberWithLong:value];

   fails  = 0;
   fails += check( "char",               nr, [NSNumber numberWithChar:value], &index);
   fails += check( "unsigned char",      nr, [NSNumber numberWithUnsignedChar:value], &index);
   fails += check( "short",              nr, [NSNumber numberWithShort:value], &index);
   fails += check( "unsigned short",     nr, [NSNumber numberWithUnsignedShort:value], &index);
   fails += check( "int",                nr, [NSNumber numberWithInt:value], &index);
   fails += check( "unsigned int",       nr, [NSNumber numberWithUnsignedInt:value], &index);
   fails += check( "long",               nr, [NSNumber numberWithLong:value], &index);
   fails += check( "unsigned long",      nr, [NSNumber numberWithUnsignedLong:value], &index);
   fails += check( "long long",          nr, [NSNumber numberWithLongLong:value], &index);
   fails += check( "unsigned long long", nr, [NSNumber numberWithUnsignedLongLong:value], &index);
   fails += check( "NSInteger",          nr, [NSNumber numberWithInteger:value], &index);
   fails += check( "NSUInteger",         nr, [NSNumber numberWithUnsignedInteger:value], &index);

   // invalid fp value > undefined behaviour
   if( value != LLONG_MAX && value != LLONG_MIN)
   {
      fails += check( "float",           nr, [NSNumber numberWithFloat:value], &index);
      fails += check( "double",          nr, [NSNumber numberWithDouble:value], &index);
#ifdef __MULLE_OBJC__
      fails += check( "long double",     nr, [NSNumber numberWithLongDouble:value], &index);
#endif
   }
   if( fails == 0)
      printf( "%ld PASSED\n\n", value);
   else
      printf( "%ld FAILED\n\n", value);
}



int  main( void)
{
   test( LONG_MIN);
   test( INT_MIN);
   test( SHRT_MIN);
   test( CHAR_MIN);
   test( NO);
   test( YES);
   test( CHAR_MAX);
   test( SHRT_MAX);
   test( INT_MAX);
   test( LONG_MAX);

   return( 0);
}
