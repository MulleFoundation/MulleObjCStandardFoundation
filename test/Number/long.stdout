1: char Number FAIL -9223372036854775808 vs. 0
2: char Number FAIL -9223372036854775808 vs. 0
3: unsigned char Number FAIL -9223372036854775808 vs. 0
4: unsigned char Number FAIL -9223372036854775808 vs. 0
5: short Number FAIL -9223372036854775808 vs. 0
6: short Number FAIL -9223372036854775808 vs. 0
7: unsigned short Number FAIL -9223372036854775808 vs. 0
8: unsigned short Number FAIL -9223372036854775808 vs. 0
9: int Number FAIL -9223372036854775808 vs. 0
10: int Number FAIL -9223372036854775808 vs. 0
11: unsigned int Number FAIL -9223372036854775808 vs. 0
12: unsigned int Number FAIL -9223372036854775808 vs. 0
15: unsigned long Number FAIL -9223372036854775808 vs. 9223372036854775808
16: unsigned long Number FAIL -9223372036854775808 vs. 9223372036854775808
19: unsigned long long Number FAIL -9223372036854775808 vs. 9223372036854775808
20: unsigned long long Number FAIL -9223372036854775808 vs. 9223372036854775808
23: NSUInteger Number FAIL -9223372036854775808 vs. 9223372036854775808
24: NSUInteger Number FAIL -9223372036854775808 vs. 9223372036854775808
-9223372036854775808 FAILED

1: char Number FAIL -2147483648 vs. 0
2: char Number FAIL -2147483648 vs. 0
3: unsigned char Number FAIL -2147483648 vs. 0
4: unsigned char Number FAIL -2147483648 vs. 0
5: short Number FAIL -2147483648 vs. 0
6: short Number FAIL -2147483648 vs. 0
7: unsigned short Number FAIL -2147483648 vs. 0
8: unsigned short Number FAIL -2147483648 vs. 0
11: unsigned int Number FAIL -2147483648 vs. 2147483648
12: unsigned int Number FAIL -2147483648 vs. 2147483648
15: unsigned long Number FAIL -2147483648 vs. 18446744071562067968
16: unsigned long Number FAIL -2147483648 vs. 18446744071562067968
19: unsigned long long Number FAIL -2147483648 vs. 18446744071562067968
20: unsigned long long Number FAIL -2147483648 vs. 18446744071562067968
23: NSUInteger Number FAIL -2147483648 vs. 18446744071562067968
24: NSUInteger Number FAIL -2147483648 vs. 18446744071562067968
-2147483648 FAILED

1: char Number FAIL -32768 vs. 0
2: char Number FAIL -32768 vs. 0
3: unsigned char Number FAIL -32768 vs. 0
4: unsigned char Number FAIL -32768 vs. 0
7: unsigned short Number FAIL -32768 vs. 32768
8: unsigned short Number FAIL -32768 vs. 32768
11: unsigned int Number FAIL -32768 vs. 4294934528
12: unsigned int Number FAIL -32768 vs. 4294934528
15: unsigned long Number FAIL -32768 vs. 18446744073709518848
16: unsigned long Number FAIL -32768 vs. 18446744073709518848
19: unsigned long long Number FAIL -32768 vs. 18446744073709518848
20: unsigned long long Number FAIL -32768 vs. 18446744073709518848
23: NSUInteger Number FAIL -32768 vs. 18446744073709518848
24: NSUInteger Number FAIL -32768 vs. 18446744073709518848
-32768 FAILED

3: unsigned char Number FAIL -128 vs. 128
4: unsigned char Number FAIL -128 vs. 128
7: unsigned short Number FAIL -128 vs. 65408
8: unsigned short Number FAIL -128 vs. 65408
11: unsigned int Number FAIL -128 vs. 4294967168
12: unsigned int Number FAIL -128 vs. 4294967168
15: unsigned long Number FAIL -128 vs. 18446744073709551488
16: unsigned long Number FAIL -128 vs. 18446744073709551488
19: unsigned long long Number FAIL -128 vs. 18446744073709551488
20: unsigned long long Number FAIL -128 vs. 18446744073709551488
23: NSUInteger Number FAIL -128 vs. 18446744073709551488
24: NSUInteger Number FAIL -128 vs. 18446744073709551488
-128 FAILED

0 PASSED

1 PASSED

127 PASSED

1: char Number FAIL 32767 vs. -1
2: char Number FAIL 32767 vs. -1
3: unsigned char Number FAIL 32767 vs. 255
4: unsigned char Number FAIL 32767 vs. 255
32767 FAILED

1: char Number FAIL 2147483647 vs. -1
2: char Number FAIL 2147483647 vs. -1
3: unsigned char Number FAIL 2147483647 vs. 255
4: unsigned char Number FAIL 2147483647 vs. 255
5: short Number FAIL 2147483647 vs. -1
6: short Number FAIL 2147483647 vs. -1
7: unsigned short Number FAIL 2147483647 vs. 65535
8: unsigned short Number FAIL 2147483647 vs. 65535
25: float Number FAIL 2147483647 vs. 2147483648
26: float Number FAIL 2147483647 vs. 2147483648
2147483647 FAILED

1: char Number FAIL 9223372036854775807 vs. -1
2: char Number FAIL 9223372036854775807 vs. -1
3: unsigned char Number FAIL 9223372036854775807 vs. 255
4: unsigned char Number FAIL 9223372036854775807 vs. 255
5: short Number FAIL 9223372036854775807 vs. -1
6: short Number FAIL 9223372036854775807 vs. -1
7: unsigned short Number FAIL 9223372036854775807 vs. 65535
8: unsigned short Number FAIL 9223372036854775807 vs. 65535
9: int Number FAIL 9223372036854775807 vs. -1
10: int Number FAIL 9223372036854775807 vs. -1
11: unsigned int Number FAIL 9223372036854775807 vs. 4294967295
12: unsigned int Number FAIL 9223372036854775807 vs. 4294967295
9223372036854775807 FAILED

